components/collator/src/comparison.rs

```diff
@@ -62,6 +71,118 @@ impl AnyQuaternaryAccumulator {
     }
 }
 
+#[inline(always)]
+fn in_inclusive_range16(i: u16, start: u16, end: u16) -> bool {
+    i.wrapping_sub(start) <= (end - start)
+}
+
+fn split_prefix_u16<'a, 'b>(
+    left: &'a [u16],
+    right: &'b [u16],
+) -> (&'a [u16], &'a [u16], &'b [u16]) {
+    let mut i = left
+        .iter()
+        .zip(right.iter())
+        .take_while(|(l, r)| l == r)
+        .count();
+    if i != 0 {
+        if let Some(&last) = left.get(i.wrapping_sub(1)) {
+            if in_inclusive_range16(last, 0xD800, 0xDBFF) {
+                i -= 1;
+            }
+            if let Some((head, left_tail)) = left.split_at_checked(i) {
+                if let Some(right_tail) = right.get(i..) {
+                    return (head, left_tail, right_tail);
+                }
+            }
+        }
+    }
+    (&[], left, right)
+}
+
+fn split_prefix_u8<'a, 'b>(left: &'a [u8], right: &'b [u8]) -> (&'a [u8], &'a [u8], &'b [u8]) {
+    let mut i = left
+        .iter()
+        .zip(right.iter())
+        .take_while(|(l, r)| l == r)
+        .count();
+    if i != 0 {
+        // Tails must not start with a UTF-8 continuation
+        // byte unless it's the first byte of the original
+        // slice.
+
+        // First, left and right differ, but since they
+        // are the same afterwards, one of them needs checking
+        // only once.
+        if let Some(right_first) = right.get(i) {
+            if (right_first & 0b1100_0000) == 0b1000_0000 {
+                i -= 1;
+            }
+        }
+        while i != 0 {
+            if let Some(left_first) = left.get(i) {
+                if (left_first & 0b1100_0000) == 0b1000_0000 {
+                    i -= 1;
+                    continue;
+                }
+            }
+            break;
+        }
+        if let Some((head, left_tail)) = left.split_at_checked(i) {
+            if let Some(right_tail) = right.get(i..) {
+                return (head, left_tail, right_tail);
+            }
+        }
+    }
+    (&[], left, right)
+}
+
+fn split_prefix<'a, 'b>(left: &'a str, right: &'b str) -> (&'a str, &'a str, &'b str) {
+    let left_bytes = left.as_bytes();
+    let right_bytes = right.as_bytes();
+    let mut i = left_bytes
+        .iter()
+        .zip(right_bytes.iter())
+        .take_while(|(l, r)| l == r)
```

Comment:
suggestion: just use `char_indices()` to avoid the continuation byte check

Reply:
`char_indices` would run UTF-8 math on all characters in the prefix, so I'm assuming that as the prefix gets longer, the code as written gets faster relative to `char_indices`. It would be good to actually measure as a follow-up, but not touching this right now to get this landed.